### PR TITLE
Overhaul documentation

### LINKED ISSUE #361: Review CLI docs
It has been suggested that our CLI docs are out of date. This issue is just a reminder to review and remedy where necessary.

### LINKED ISSUE #447: Review and update docs
Some information is out of date or in the wrong place, in particular a lot of detailed technical explanation has made into "how-tos" when it should probably be in "explanations"

## Acceptance Criteria
- The docs have been reviewed
- Information that is no longer true has been removed
- Information has been refactored into its proper place in the structure

### LINKED ISSUE #671: Document Dev Environment
## Acceptance Criteria
- Docs exist on how to run blueapi locally with
  - Message bus
  - Custom plans/devices

### PR DESCRIPTION
**Pull request by callumforrester**
Fixes #447 
Fixes #361 
Fixes #671

### CODE REVIEW

File: docs/explanations/architecture.md
Comment (DiamondJoseph) [inline docs/explanations/architecture.md:14]:
```suggestion
Above are the main components of blueapi. The main process houses the REST API and manages the subprocess, which wraps the `RunEngine`, devices and external connections.
```
Do we want to be general or specific?

File: docs/explanations/extension-code.md
Comment (DiamondJoseph) [inline docs/explanations/extension-code.md:3]: I'd move the device/dodal part first, so that it's clear that devices belong in dodal, rather than someone adding devices to their repo then realising it's meant to be in dodal.

File: docs/explanations/extension-code.md
Comment (DiamondJoseph) [inline docs/explanations/extension-code.md:17]: new devices = new device classes?

This is mostly true, get devices into dodal, then plan to move to ophyd-async. Let's treat mostly true as true for the purposes of introductory docs.

Should be expanded to explain that beamlines should also be added there, with a link to dodal how to make a new beamline.

File: docs/explanations/extension-code.md
Comment (DiamondJoseph) [inline docs/explanations/extension-code.md:9]: Could: an example, like i22?

File: docs/explanations/lifecycle.md
Comment (DiamondJoseph) [inline docs/explanations/lifecycle.md:16]: Is this change going in prior to deprecating inject or after?

Comment (DiamondJoseph) [reply docs/explanations/lifecycle.md:16]: Update to >=3.10 typing (List->list, Union[x, y] -> x | y, Optional[x] -> x | None)

File: docs/explanations/lifecycle.md
Comment (DiamondJoseph) [inline docs/explanations/lifecycle.md:120]: >* Changes to any device statuses running within a plan (e.g. when a motor changes position)

should be `the` plan as well

File: docs/explanations/type_validators.md
Comment (DiamondJoseph) [inline docs/explanations/type_validators.md:14]: Is this repeating what is explained in plans.md?

File: docs/explanations/type_validators.md
Comment (DiamondJoseph) [inline docs/explanations/type_validators.md:89]: Python 3.10ify

File: docs/how-to/configure-app.md
Comment (DiamondJoseph) [inline docs/how-to/configure-app.md:4]: memo: this is about to get changed by changes to overriding config values from env vars etc, make sure to update it then

File: docs/how-to/write-devices.md
Comment (DiamondJoseph) [inline docs/how-to/write-devices.md:25]: Should: defer to dodal's docs on writing device classes and beamlines, and ensure they are up to date?

File: docs/how-to/write-plans.md
Comment (DiamondJoseph) [inline docs/how-to/write-plans.md:92]: Should: replacing this form of docstring with using Annotated/Field arguments so that the values propagate into json schema?

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]
How much is this just re-formatting the docs to pass current linting/build and how much is it intending to update the docs? There's several things that we're actively moving away from that aren't being changed, but also more changes than just updating the formatting.

File: docs/explanations/architecture.md
Comment (callumforrester) [reply docs/explanations/architecture.md:14]: I made this a bit more general because I didn't think a list of major classes and their purposes was _that_ useful and was prone to being out of date. It may be more useful when blueapi is more stable.

File: docs/explanations/extension-code.md
Comment (callumforrester) [reply docs/explanations/extension-code.md:17]: This is a problem in general, some of these docs belong in dodal with links from blueapi. I think we can and should do that migration gradually though. I'm going to advocate for removing this sentence and making an issue in dodal for adding a page/pages explaining how to:
- Make a new device
- Add a device to a beamline
- Make a new beamline

File: docs/explanations/lifecycle.md
Comment (callumforrester) [reply docs/explanations/lifecycle.md:16]: I am going to make a separate issue for moving all of the code snippets to Python files and checking that they still work and lint properly at docs build time, similar to ophyd-async.

Comment (callumforrester) [reply docs/explanations/lifecycle.md:16]: #677

File: docs/explanations/extension-code.md
Comment (DiamondJoseph) [reply docs/explanations/extension-code.md:17]: The reason I asked is there is already docs in dodal for at least part of that
https://diamondlightsource.github.io/dodal/main/how-to/create-beamline.html
https://diamondlightsource.github.io/dodal/main/reference/device-standards.html (?)

File: docs/explanations/type_validators.md
Comment (callumforrester) [reply docs/explanations/type_validators.md:14]: Probably does need a bit of consolidation, my read is that plans.md explains which parts of blueapi plans are special and why, and type_validators.md explains the very technical implementation of plan parameter parsing.

File: docs/explanations/type_validators.md
Comment (callumforrester) [reply docs/explanations/type_validators.md:89]: #677

File: docs/how-to/write-devices.md
Comment (callumforrester) [reply docs/how-to/write-devices.md:25]: Agreed, see other discussions, I think we should migrate some of this to dodal as a separate PR

File: docs/how-to/write-plans.md
Comment (callumforrester) [reply docs/how-to/write-plans.md:92]: Oh, I did not realise they didn't propagate, that's quite bad. I was hoping to make the plans as vanilla-bluesky as we could, so vanilla python function docstrings are definitely desirable...

Comment (DiamondJoseph) [reply docs/how-to/write-plans.md:92]: Lemme get a couple of example plans into blueapi and compare the json schemas

**Comment by callumforrester** [comment]
@DiamondJoseph regarding the fact that some of this is going out of date "soon": I have been bitten too many times by writing code and docs in anticipation of changes that are due "in a week or so" and end up taking 6 months. I'm strongly against putting anything in here that is not true in `main` right now. 

Ideally docs should be updated in the same PR as the change they reference, second-best option is after the PR is merged, worst-best is before the PR is merged.

File: docs/explanations/extension-code.md
Comment (callumforrester) [reply docs/explanations/extension-code.md:17]: Ah cool! I'll reference those

File: docs/how-to/write-plans.md
Comment (DiamondJoseph) [reply docs/how-to/write-plans.md:92]:
```python
@attach_data_session_metadata_decorator()
@validate_call(config={"arbitrary_types_allowed": True})
def linkam_plan_1(
    trajectory: LinkamTrajectory,
    linkam: Linkam3 = LINKAM,
    shutter_time: float = 0.04,
) -> MsgGenerator:
    """
    Follow a trajectory in temperature, collecting a number of frames either at equally
    spaced positions or while continually scanning. e.g. for 2 segments, the first
    stepped and the 2nd flown:\n
    trajectory start   v             v final segment stop\n
                       \\           /\n
       stepped segment__\\__       /\n
                           \\     /  flown segment\n
           1st segment stop \\__ /\n
        exposures:    xx  xx  xx   1/N seconds

    Args:
        trajectory: Trajectory for the scan to follow.
        linkam: Temperature controller
        shutter_time: Time allowed (s) for opening shutter before triggering detectors.

    Returns:
            MsgGenerator: Plan

    Yields:
            Iterator[MsgGenerator]: Bluesky messages
    """
    ...

@attach_data_session_metadata_decorator()
@validate_call(config={"arbitrary_types_allowed": True})
def linkam_plan_2(
    trajectory: Annotated[
        LinkamTrajectory, Field(description="Trajectory for the scan to follow.")
    ],
    linkam: Annotated[Linkam3, Field(LINKAM, description="Temperature controller.")],
    shutter_time: Annotated[
        float,
        Field(
            0.04,
            description="Time allowed for opening shutter before triggering detectors.",
            json_schema_extra={"units": "s"},
        ),
    ],
) -> MsgGenerator:
    """
    Follow a trajectory in temperature, collecting a number of frames either at equally
    spaced positions or while continually scanning. e.g. for 2 segments, the first
    stepped and the 2nd flown:\n
    trajectory start   v             v final segment stop\n
                       \\           /\n
       stepped segment__\\__       /\n
                           \\     /  flown segment\n
           1st segment stop \\__ /\n
        exposures:    xx  xx  xx   1/N seconds
    """
    ...

@attach_data_session_metadata_decorator()
@validate_call(config={"arbitrary_types_allowed": True})
def linkam_plan_3(
    trajectory: LinkamTrajectory = Field(
        description="Trajectory for the scan to follow."
    ),
    linkam: Linkam3 = Field(LINKAM, description="Temperature controller."),
    shutter_time: float = Field(
        0.04,
        description="Time allowed for opening shutter before triggering detectors.",
        json_schema_extra={"units": "s"},
    ),
) -> MsgGenerator:
    """
    Follow a trajectory in temperature, collecting a number of frames either at equally
    spaced positions or while continually scanning. e.g. for 2 segments, the first
    stepped and the 2nd flown:\n
    trajectory start   v             v final segment stop\n
                       \\           /\n
       stepped segment__\\__       /\n
                           \\     /  flown segment\n
           1st segment stop \\__ /\n
        exposures:    xx  xx  xx   1/N seconds
    """
    ...
```

[linkam_plan_1.json](https://github.com/user-attachments/files/17436590/linkam_plan_1.json)
[linkam_plan_2.json](https://github.com/user-attachments/files/17436591/linkam_plan_2.json)
[linkam_plan_3.json](https://github.com/user-attachments/files/17436656/linkam_plan_3.json)

I've just gone through all that just to prove they're basically identical and the Pydantic model doesn't actually get the information about the Fields :upside_down_face:

Comment (callumforrester) [reply docs/how-to/write-plans.md:92]: I think we can continue this discussion in #644

File: docs/how-to/write-plans.md
Comment (callumforrester) [inline docs/how-to/write-plans.md:51]: `inject` is now deprecated, will update these before merging

File: docs/explanations/architecture.md
Comment (DiamondJoseph) [reply docs/explanations/architecture.md:14]: Lost in translation- I'm suggesting a change to make it slightly more generic but hopefully still make sense?

File: docs/tutorials/run-bus.md
Comment (callumforrester) [inline docs/tutorials/run-bus.md:16]: Remove per discussion with @DominicOram

File: docs/explanations/lifecycle.md
Comment (DiamondJoseph) [inline docs/explanations/lifecycle.md:12]: Should:

```suggestion
from bluesky.protocols import Readable
from bluesky.utils import MsgGenerator
```

It would be a nit, but we've seen how much of an issue this line has caused!!

File: docs/explanations/lifecycle.md
Comment (DiamondJoseph) [inline docs/explanations/lifecycle.md:116]: nit

```suggestion
* Changes to any device statuses running within the plan (e.g. when a motor changes position)
```

File: docs/explanations/plans.md
Comment (DiamondJoseph) [inline docs/explanations/plans.md:14]:
```suggestion
Allowed argument types for Pydantic BaseModels include the primitives, types that extend `BaseModel` and `dict`'s, `list`'s  and other `sequence`'s of supported types. Blueapi will deserialise these types from JSON, so `dict`'s must use `str` keys.
```

nit

File: docs/explanations/type_validators.md
Comment (DiamondJoseph) [reply docs/explanations/type_validators.md:14]: I think this is just repeating what is in lifecycle.md. It even includes the same json example, I'd really like to see one removed or the difference between the two explained.

File: docs/how-to/write-devices.md
Comment (DiamondJoseph) [reply docs/how-to/write-devices.md:25]: dodal already has docs on creating a beamline and making a device class. 

https://diamondlightsource.github.io/dodal/main/how-to/create-beamline.html
and
https://diamondlightsource.github.io/dodal/main/how-to/make-new-ophyd-async-device.html

We should defer to these and fix those pages. The former should be updated to use device_factory and the second should be fixed as the Mermaid docs do not render responsively and should explain how to make a Device class.

File: docs/how-to/write-devices.md
Comment (DiamondJoseph) [inline docs/how-to/write-devices.md:33]: must:
https://github.com/DiamondLightSource/dodal/blob/a3cf0ecf432dc7ad2a62c2583e90d452154f9ef5/src/dodal/plan_stubs/data_session.py#L43
```suggestion
Dodal defines a decorator, `@attach_data_session_metadata_decorator()`, for configuring `ophyd-async` detectors to write data to a common location. 
```

could: explain that it requires the get_path_provider from dodal is used as constructor arg?

File: docs/how-to/write-plans.md
Comment (DiamondJoseph) [inline docs/how-to/write-plans.md:46]: I think we need to tighten up Mapping -> dict to behave well with bp and bps usage.

**Review by DiamondJoseph — CHANGES_REQUESTED** [review]

File: docs/how-to/write-plans.md
Comment (callumforrester) [reply docs/how-to/write-plans.md:46]: Sorry, not sure I follow, what issues are you seeing?

Comment (DiamondJoseph) [reply docs/how-to/write-plans.md:46]: bluesky.plans have their equivalent md argument (which we should use when possible) is `CustomPlanMetadata = dict[str, Any]`: we therefore cannot pass a `Mapping[str, Any]` down to it: `Mapping` does not allow assignment, where `dict` does. We also sometimes need to mutate metadata in a "sub"plan before passing it to a "super"plan.

File: docs/explanations/type_validators.md
Comment (callumforrester) [reply docs/explanations/type_validators.md:14]: So the idea is that `lifecycle.md` is a general overview of how blueapi stores, retrieves and runs plans whereas this is a deep dive into the special handling for substituting devices for strings. Do you have any suggestions as to what we can shuffle around to make that clearer?

File: docs/how-to/write-devices.md
Comment (callumforrester) [reply docs/how-to/write-devices.md:25]: I agree, but I would rather merge this now and raise separate issues for fixing those pages and consolidating, otherwise this PR will languish for a long time and the published docs (that everyone reads) will continue to contain mistakes that throw people off. Alternatively, I can remove this file entirely from the PR and make issues in dodal, what do you think?

File: docs/how-to/write-plans.md
Comment (DiamondJoseph) [reply docs/how-to/write-plans.md:46]: This very plan would fail type checking for the issue. We can quite neatly then have the required parts of the plan be from bluesky.utils:

```python
from bluesky.utils import CustomPlanMetadata, MsgGenerator

def foo(metadata: CustomPlanMetadata | None = None) -> MsgGenerator:
    yield from {}
```

Comment (callumforrester) [reply docs/how-to/write-plans.md:46]: I'm happy to change it, please feel free to read or ignore the small rant below.

The reason it's `dict` is because it's mutated upstream:

```python
def my_plan(params: ..., md: dict[str, Any]):
    ...
    md["plan_name"] = my_plan.__name__
    yield from ...
```

It could instead be kept immutable:

```python
def my_plan(params: ..., md: Mapping[str, Any]):
    ...
    md = {**md, **{"plan_name": my_plan.__name__}}
    yield from ...
```

Immutable = good, so I would prefer we change upstream, but the verbosity of that solution may be unpalletable.

File: docs/how-to/write-devices.md
Comment (callumforrester) [reply docs/how-to/write-devices.md:25]: https://github.com/DiamondLightSource/dodal/issues/933

File: docs/how-to/write-plans.md
Comment (DiamondJoseph) [inline docs/how-to/write-plans.md:25]: nit: the plan linting in dodal enforces that plans have exactly this argument `metadata: dict[str, Any] | None = None` - would be nice to have this example compliant with that.

Comment (DiamondJoseph) [reply docs/how-to/write-plans.md:25]: See also line 43 of this file which says that plans should have this arg...

**Review by DiamondJoseph — APPROVED** [review]

**Pull request merged**
